Code/RDGeneral/ConcurrentQueue.h

```diff
@@ -0,0 +1,102 @@
+#ifdef RDK_THREADSAFE_SSS
+#ifndef CONCURRENT_QUEUE
+#define CONCURRENT_QUEUE
+#include <condition_variable>
+#include <queue>
+#include <thread>
+
+namespace RDKit {
+template <class E>
+class ConcurrentQueue {
+ private:
+  size_t capacity;
+  bool done;
+  std::mutex lock;
+  //! we need two condition variables to establish communication
+  //! between popping (consumer threads) and pushing (producer threads)
+  std::condition_variable goPush, goPop;
+  std::queue<E> q;
+
+ public:
+  ConcurrentQueue<E>(size_t capacity) : capacity(capacity), done(false) {}
+  //! tries to push an element into the queue if it is not full without
+  //! modifying the variable element
+  void push(const E& element);
+
+  //! tries to pop an element from the queue if it is not empty and not done
+  //! the boolean value indicates the whether popping is successful
```

Comment:
It would be good to document here why the pop would be unsuccessful (in this case I think it's because the producer has marked the queue as done.